yktracec/src/jitmodbuilder.cc

```diff
@@ -1744,9 +1722,17 @@ class JITModBuilder {
     // Carry on constructing the trace module in the success block.
     Builder.SetInsertPoint(SuccBB);
 
-    // Update the VMap so that the promoted value is now a constant.
-    VMap[CI] = PConstLL;
-    PromoteMap[CI] = JITPromoteVar;
+    // You might think that we need to do: `VMap[CI] = PConstLL` so as to use
+    // the constant in place of the promoted value from here onward.
+    //
+    // As it happens, the guard we just inserted already gives LLVM everything
+    // it needs to promote the value for us.
+    //
+    // If we trace a `yk_promote(x)` when `x==5`, then the trace compiler
+    // makes a guard that says "if x!=5 then goto the guard failure block,
+    // otherwise goto the guard success block". LLVM can then easily infer
```

Comment:
In hindsight, I wonder if we should replace "easily" with "often (with sufficient AOT and JIT LLVM passes)", because we only noticed this behaviour when turning on the AOT `instcombine` pass.

Reply:
I would probably say "LLVM is then more likely to be able to infer that...".

Reply:
how about "llvm is likely to be able to...".

Wasn't sure about "more likely" as without the guard it would be incorrect for LLVM to make that deduction at all.

Reply:
*Technically* nothing stops LLVM putting in a speculation (i.e. compiling two code paths). At some point, though, we're in angels-on-a-pin territory!

Reply:
What I mean is that without the guard, nothing is known about the promoted value, so it would be miscompilation for llvm to assume that it is any constant value.  So "more likely" makes no sense, as without the guard it should be "impossible" for it to make any deduction of its value.

Reply:
That's pretty convoluted and stating that doesn't help this explanation.

I could walk backwards everywhere I go, but I don't :P

Reply:
Yes, but there is a mildly interesting point: "miscompilation" is a more subtle concept than it first appears.

All of this is why I tend to prefer something like "more likely to be able to infer that" as it covers a multitude of possibilities.

Reply:
I find "more likely" confusing -- is there something less confusing that we can agree on?

Reply:
I'm unsure what could be less confusing; or, at least, "less confusing" might end up being "less accurate"? Ultimately, this is a relative concept: LLVM can *already* infer this property if it so chooses. Promotion just makes it (much) more likely that it infers this property, but it's still not a guarantee that it does so.

Reply:
OK.